Fixed #31246 -- Fixed locking models in QuerySet.select_for_update(of=()) for related fields and parent link fields with multi-table inheritance.

abhijeetviswa opened the pull request:
When a related model (with parents) is specified in `of` for select_for_update,
the `of` clause of the generated query references the field's parent model
instead of the field's model.

abhijeetviswa commented on django/db/models/sql/compiler.py:
```diff
@@ -997,19 +997,21 @@ def _get_field_choices():
                     for klass_info in klass_info.get('related_klass_infos', [])
                 )
+
+        def _get_col_from_model(klass_info):
```
Thanks. I'll fix these.

felixxm commented:
@abhijeetviswa Thanks, please include tests and fix for multi-level inheritance, and add release notes for Django 2.2.10.

abhijeetviswa commented:
@felixxm I'm not sure about the wording for the fix in the release notes. Could you please check it out?

Also, I'm not sure what more tests to add. I've modified `test_for_update_sql_model_inheritance_nested_ptr_generated_of` to test for related fields with inheritance and added `test_for_update_sql_multilevel_model_intehritance_ptr_generated_of` to test for multi-level inheritance.

felixxm commented on tests/select_for_update/tests.py:
```diff
@@ -148,15 +150,30 @@ def test_for_update_sql_model_inheritance_ptr_generated_of(self):
         with transaction.atomic(), CaptureQueriesContext(connection) as ctx:
             list(EUCity.objects.select_related('country').select_for_update(
-                of=('self', 'country__country_ptr',),
+                of=('self', 'country', 'country__country_ptr',),
```
This is not related with using nested `_ptr`, so we should add a new test for this bugfix.

felixxm commented:
buildbot, test on oracle.

carltongibson commented:
OK, looks good. 😅 

@felixxm Oracle failure seems unrelated — test didn't fail — but...

felixxm commented:
Thanks for checking. Yes, Oracle failure is not related.
